File: test/product_api_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;

import 'package:calcoricher/models/daily_question.dart';
import 'package:calcoricher/models/feed_item.dart';
import 'package:calcoricher/models/receipt_model.dart';
import 'package:calcoricher/models/room_model.dart';
import 'package:calcoricher/services/product_api.dart';

void main() {
  test('ProductApi maps auth, guardrails, and purchase responses', () async {
    final requests = <http.BaseRequest>[];
    final client = _FakeClient((request) async {
      requests.add(request);
      final body = request is http.Request && request.body.isNotEmpty
          ? jsonDecode(request.body)
          : null;

      if (request.method == 'GET' && request.url.path == '/api/guardrails') {
        return _json({
          'dailySpendLimitMinor': 10000,
          'currencyCode': 'chf',
          'currencySymbol': 'CHF',
          'satireDisclosure': 'This is satire.',
          'helpUrl': 'mailto:help@example.test',
          'refundUrl': 'mailto:refund@example.test',
          'priceLadder': [
            {'unlock': 1, 'amountMinor': 100, 'label': 'CHF 1.00'},
            {'unlock': 2, 'amountMinor': 200, 'label': 'CHF 2.00'},
          ],
        });
      }

      if (request.method == 'POST' &&
          request.url.path == '/api/auth/register') {
        expect(body, {
          'username': 'Ada',
          'email': 'ada@example.test',
          'password': 'pass1234',
        });
        return _json({
          'token': 'token_123',
          'user': {
            'id': 'user_1',
            'username': 'Ada',
            'email': 'ada@example.test',
            'totalSpentMinor': 0,
            'unlockedResultsCount': 0,
            'highestUnlockMinor': 0,
            'currentResultPriceMinor': 100,
            'badges': [],
            'titles': [],
            'receipts': [],
          },
        }, status: 201);
      }

      if (request.method == 'POST' && request.url.path == '/api/purchases') {
        expect(request.headers['authorization'], 'Bearer token_123');
        expect(body['expression'], '17 * 3');
        expect(body['context']['roomCode'], 'ABC123');
        return _json({
          'purchase': {
            'id': 'purchase_1',
            'expression': '17 * 3',
            'result': '51',
            'amountMinor': 400,
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'receipt': {
            'id': 'receipt_1',
            'purchaseId': 'purchase_1',
            'expression': '17 * 3',
            'result': '51',
            'amountMinor': 400,
            'rank': 1,
            'shareText': 'I paid CHF 4.00 for this answer.',
            'imageUrl': '/api/receipts/receipt_1.svg',
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'feedItem': {
            'id': 'feed_1',
            'purchaseId': 'purchase_1',
            'receiptId': 'receipt_1',
            'by': 'Ada',
            'userId': 'user_1',
            'expression': '17 * 3',
            'result': '51',
            'amountMinor': 400,
            'shareText': 'I paid CHF 4.00 for this answer.',
            'roomCode': 'ABC123',
            'challengeSlug': 'streamer-night',
            'charityCampaignId': 'math-relief',
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'user': {
            'id': 'user_1',
            'username': 'Ada',
            'email': 'ada@example.test',
            'totalSpentMinor': 400,
            'unlockedResultsCount': 1,
            'highestUnlockMinor': 400,
            'currentResultPriceMinor': 200,
            'badges': [
              {'id': 'first-reveal', 'title': 'First Reveal'},
            ],
            'titles': ['Receipt Collector'],
            'receipts': [],
          },
        }, status: 201);
      }

      throw StateError('Unexpected request ${request.method} ${request.url}');
    });

    final api = ProductApi(
      baseUri: Uri.parse('https://api.example.test'),
      client: client,
    );

    final guardrails = await api.guardrails();
    expect(guardrails.dailySpendLimitMinor, 10000);
    expect(guardrails.priceLadder.first.label, 'CHF 1.00');

    final session = await api.register(
      username: 'Ada',
      email: 'ada@example.test',
      password: 'pass1234',
    );
    expect(session.token, 'token_123');
    expect(session.user.username, 'Ada');

    final purchase = await api.createPurchase(
      token: session.token,
      expression: '17 * 3',
      result: '51',
      amountMinor: 400,
      context: const PurchaseContext(
        roomCode: 'ABC123',
        challengeSlug: 'streamer-night',
        dailyQuestionDate: '2026-06-21',
        charityCampaignId: 'math-relief',
      ),
    );
    expect(purchase.receipt.shareText, 'I paid CHF 4.00 for this answer.');
    expect(purchase.receipt.rank, 1);
    expect(purchase.feedItem.by, 'Ada');
    expect(purchase.user.totalSpentMinor, 400);
    expect(requests.length, 3);
  });

  test('models parse daily questions, rooms, receipts, and feed items', () {
    final daily = DailyQuestion.fromJson({
      'date': '2026-06-21',
      'expression': '17 * 3',
      'title': 'Daily Rich Question',
    });
    expect(daily.expression, '17 * 3');

    final room = RoomModel.fromJson({
      'id': 'room_1',
      'code': 'ABC123',
      'title': 'Sunday Rich Room',
      'ownerId': 'user_1',
      'members': ['user_1', 'user_2'],
      'createdAt': '2026-06-21T12:00:00.000Z',
    });
    expect(room.members, contains('user_2'));

    final receipt = ReceiptModel.fromJson({
      'id': 'receipt_1',
      'purchaseId': 'purchase_1',
      'expression': '2 + 2',
      'result': '4',
      'amountMinor': 100,
      'rank': 7,
      'shareText': 'I paid CHF 1.00 for this answer.',
      'imageUrl': '/api/receipts/receipt_1.svg',
      'timestamp': '2026-06-21T12:00:00.000Z',
    });
    expect(receipt.imageUrl, '/api/receipts/receipt_1.svg');
    expect(receipt.rank, 7);

    final feed = FeedItem.fromJson({
      'id': 'feed_1',
      'purchaseId': 'purchase_1',
      'receiptId': 'receipt_1',
      'by': 'Lord Spreadsheet',
      'userId': 'user_1',
      'expression': '2 + 2',
      'result': '4',
      'amountMinor': 100,
      'shareText': 'I paid CHF 1.00 for this answer.',
      'roomCode': null,
      'challengeSlug': null,
      'charityCampaignId': null,
      'timestamp': '2026-06-21T12:00:00.000Z',
    });
    expect(feed.by, 'Lord Spreadsheet');
  });

  test('ProductApi maps competition routes and duration-aware purchases',
      () async {
    final client = _FakeClient((request) async {
      final body = request is http.Request && request.body.isNotEmpty
          ? jsonDecode(request.body)
          : null;

      if (request.method == 'POST' && request.url.path == '/api/purchases') {
        expect(body['durationMs'], 1200);
        return _json({
          'purchase': {
            'id': 'purchase_1',
            'expression': '2 + 2',
            'result': '4',
            'amountMinor': 500,
            'durationMs': 1200,
            'ridiculousScore': 17,
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'receipt': {
            'id': 'receipt_1',
            'purchaseId': 'purchase_1',
            'expression': '2 + 2',
            'result': '4',
            'amountMinor': 500,
            'rank': 2,
            'shareText': 'I paid CHF 5.00 for this answer.',
            'imageUrl': '/api/receipts/receipt_1.svg',
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'feedItem': {
            'id': 'feed_1',
            'purchaseId': 'purchase_1',
            'receiptId': 'receipt_1',
            'by': 'Bob',
            'userId': 'user_2',
            'expression': '2 + 2',
            'result': '4',
            'amountMinor': 500,
            'shareText': 'I paid CHF 5.00 for this answer.',
            'roomCode': 'ABC123',
            'challengeSlug': 'streamer-night',
            'charityCampaignId': null,
            'timestamp': '2026-06-21T12:00:00.000Z',
          },
          'user': {
            'id': 'user_2',
            'username': 'Bob',
            'email': 'bob@example.test',
            'totalSpentMinor': 500,
            'unlockedResultsCount': 1,
            'highestUnlockMinor': 500,
            'currentResultPriceMinor': 200,
            'badges': [],
            'titles': [],
            'receipts': [],
          },
        }, status: 201);
      }

      if (request.method == 'GET' &&
          request.url.path == '/api/rooms/ABC123/competition') {
        return _json({'leaders': _competitionJson()});
      }

      if (request.method == 'GET' &&
          request.url.path == '/api/challenges/streamer-night/competition') {
        return _json({'leaders': _competitionJson()});
      }

      if (request.method == 'GET' &&
          request.url.path == '/api/weekly/2026-W25/leaderboard') {
        return _json({
          'weekKey': '2026-W25',
          'users': [_leaderJson('user_1', 'Ada', 900)],
          'leaders': _competitionJson(),
        });
      }

      throw StateError('Unexpected request ${request.method} ${request.url}');
    });

    final api = ProductApi(
      baseUri: Uri.parse('https://api.example.test'),
      client: client,
    );

    final purchase = await api.createPurchase(
      token: 'token_123',
      expression: '2 + 2',
      result: '4',
      amountMinor: 500,
      durationMs: 1200,
      context: const PurchaseContext(roomCode: 'ABC123'),
    );
    expect(purchase.purchase.durationMs, 1200);
    expect(purchase.purchase.ridiculousScore, 17);

    final room = await api.roomCompetition('ABC123');
    expect(room.spent.first.username, 'Bob');
    expect(room.fastest.first.fastestRevealMs, 1200);

    final challenge = await api.challengeCompetition('streamer-night');
    expect(challenge.highestUnlock.first.username, 'Ada');

    final weekly = await api.weeklyLeaderboard('2026-W25');
    expect(weekly.single.username, 'Ada');
  });
}

class _FakeClient extends http.BaseClient {
  final Future<http.Response> Function(http.BaseRequest request) handler;

  _FakeClient(this.handler);

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    final response = await handler(request);
    return http.StreamedResponse(
      Stream.value(response.bodyBytes),
      response.statusCode,
      headers: response.headers,
      request: request,
    );
  }
}

http.Response _json(Map<String, Object?> body, {int status = 200}) {
  return http.Response(
    jsonEncode(body),
    status,
    headers: {'content-type': 'application/json'},
  );
}

Map<String, Object?> _competitionJson() => {
      'spent': [_leaderJson('user_2', 'Bob', 1000)],
      'highestUnlock': [_leaderJson('user_1', 'Ada', 900)],
      'ridiculous': [
        {
          ..._leaderJson('user_1', 'Ada', 900),
          'ridiculousScore': 120,
        }
      ],
      'fastest': [
        {
          ..._leaderJson('user_2', 'Bob', 1000),
          'fastestRevealMs': 1200,
        }
      ],
    };

Map<String, Object?> _leaderJson(String id, String username, int spent) => {
      'id': id,
      'username': username,
      'totalSpentMinor': spent,
      'unlockedResultsCount': 1,
      'highestUnlockMinor': spent,
      'ridiculousScore': 17,
      'fastestRevealMs': null,
    };

File: server/test/product-api.test.js
const assert = require('node:assert/strict');
const { mkdtemp, rm } = require('node:fs/promises');
const { tmpdir } = require('node:os');
const path = require('node:path');
const { test } = require('node:test');

process.env.STRIPE_SECRET_KEY = process.env.STRIPE_SECRET_KEY || 'sk_test_local';
process.env.CHECKOUT_SUCCESS_URL =
  process.env.CHECKOUT_SUCCESS_URL || 'https://example.test/success';
process.env.CHECKOUT_CANCEL_URL =
  process.env.CHECKOUT_CANCEL_URL || 'https://example.test/cancel';
process.env.DAILY_SPEND_LIMIT_MINOR = '10000';

const { createApp } = require('../index');

async function withServer(fn, options = {}) {
  const dataDir = await mkdtemp(path.join(tmpdir(), 'calcoricher-api-'));
  const app = createApp({
    dataDir,
    now: options.now || (() => new Date('2026-06-21T12:00:00.000Z')),
  });
  const server = await new Promise((resolve) => {
    const listening = app.listen(0, '127.0.0.1', () => resolve(listening));
  });
  const baseUrl = `http://127.0.0.1:${server.address().port}`;

  try {
    await fn({
      dataDir,
      request: (method, url, body, token) => request(baseUrl, method, url, body, token),
    });
  } finally {
    await new Promise((resolve) => server.close(resolve));
    await rm(dataDir, { recursive: true, force: true });
  }
}

async function request(baseUrl, method, url, body, token) {
  const headers = { accept: 'application/json' };
  if (body !== undefined) headers['content-type'] = 'application/json';
  if (token) headers.authorization = `Bearer ${token}`;

  const response = await fetch(`${baseUrl}${url}`, {
    method,
    headers,
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const contentType = response.headers.get('content-type') || '';
  const payload = contentType.includes('application/json')
    ? await response.json()
    : await response.text();

  return { status: response.status, headers: response.headers, body: payload };
}

test('supports the core global product loop', async () => {
  await withServer(async ({ request }) => {
    const guardrails = await request('GET', '/api/guardrails');
    assert.equal(guardrails.status, 200);
    assert.equal(guardrails.body.dailySpendLimitMinor, 10000);
    assert.match(guardrails.body.satireDisclosure, /satire/i);
    assert.ok(guardrails.body.priceLadder.length >= 6);

    const daily = await request('GET', '/api/daily-question');
    assert.equal(daily.status, 200);
    assert.equal(daily.body.date, '2026-06-21');
    assert.match(daily.body.expression, /\d/);

    const ada = await request('POST', '/api/auth/register', {
      username: 'Ada Gold',
      email: 'ada@example.test',
      password: 'pass1234',
    });
    assert.equal(ada.status, 201);
    assert.ok(ada.body.token);
    assert.equal(ada.body.user.username, 'Ada Gold');
    assert.equal(ada.body.user.totalSpentMinor, 0);

    const login = await request('POST', '/api/auth/login', {
      email: 'ada@example.test',
      password: 'pass1234',
    });
    assert.equal(login.status, 200);
    assert.equal(login.body.user.id, ada.body.user.id);

    const room = await request(
      'POST',
      '/api/rooms',
      { title: 'Sunday Rich Room' },
      ada.body.token,
    );
    assert.equal(room.status, 201);
    assert.match(room.body.room.code, /^[A-Z0-9]{6}$/);

    const challenge = await request(
      'POST',
      '/api/challenges',
      { slug: 'streamer-night', title: 'Streamer Night' },
      ada.body.token,
    );
    assert.equal(challenge.status, 201);
    assert.equal(challenge.body.challenge.slug, 'streamer-night');

    const purchase = await request(
      'POST',
      '/api/purchases',
      {
        expression: '17 * 3',
        result: '51',
        amountMinor: 400,
        context: {
          roomCode: room.body.room.code,
          challengeSlug: 'streamer-night',
          dailyQuestionDate: daily.body.date,
          charityCampaignId: 'math-relief',
        },
      },
      ada.body.token,
    );
    assert.equal(purchase.status, 201);
    assert.equal(purchase.body.purchase.expression, '17 * 3');
    assert.equal(purchase.body.user.totalSpentMinor, 400);
    assert.equal(purchase.body.receipt.shareText, 'I paid CHF 4.00 for this answer.');
    assert.equal(purchase.body.receipt.rank, 1);
    assert.match(purchase.body.receipt.imageUrl, /^\/api\/receipts\//);
    assert.equal(purchase.body.feedItem.by, 'Ada Gold');
    assert.equal(purchase.body.feedItem.charityCampaignId, 'math-relief');

    const receipt = await request('GET', purchase.body.receipt.imageUrl);
    assert.equal(receipt.status, 200);
    assert.match(receipt.headers.get('content-type'), /^image\/svg\+xml/);
    assert.match(receipt.body, /I paid CHF 4\.00 for this answer/);
    assert.match(receipt.body, /Rank #1/);

    const leaderboard = await request('GET', '/api/leaderboard');
    assert.equal(leaderboard.status, 200);
    assert.equal(leaderboard.body.users[0].username, 'Ada Gold');
    assert.equal(leaderboard.body.users[0].totalSpentMinor, 400);

    const roomLeaderboard = await request('GET', `/api/rooms/${room.body.room.code}/leaderboard`);
    assert.equal(roomLeaderboard.status, 200);
    assert.equal(roomLeaderboard.body.users[0].username, 'Ada Gold');

    const challengeLeaderboard = await request(
      'GET',
      '/api/challenges/streamer-night/leaderboard',
    );
    assert.equal(challengeLeaderboard.status, 200);
    assert.equal(challengeLeaderboard.body.users[0].username, 'Ada Gold');

    const feed = await request('GET', '/api/feed');
    assert.equal(feed.status, 200);
    assert.equal(feed.body.items[0].expression, '17 * 3');
    assert.equal(feed.body.items[0].amountMinor, 400);

    const profile = await request('GET', `/api/users/${ada.body.user.id}`);
    assert.equal(profile.status, 200);
    assert.equal(profile.body.user.receipts.length, 1);
    assert.ok(profile.body.user.badges.some((badge) => badge.id === 'first-reveal'));
    assert.ok(profile.body.user.titles.includes('Receipt Collector'));

    const dailyLeaderboard = await request('GET', `/api/daily-question/${daily.body.date}/leaderboard`);
    assert.equal(dailyLeaderboard.status, 200);
    assert.equal(dailyLeaderboard.body.users[0].username, 'Ada Gold');
  });
});

test('supports comments, moderation, and account deletion', async () => {
  await withServer(async ({ request }) => {
    const admin = await request('POST', '/api/auth/register', {
      username: 'Admin',
      email: 'max.alberucci@gmail.com',
      password: 'pass1234',
    });
    const ada = await request('POST', '/api/auth/register', {
      username: 'Ada',
      email: 'ada@example.test',
      password: 'pass1234',
    });
    const bob = await request('POST', '/api/auth/register', {
      username: 'Bob',
      email: 'bob@example.test',
      password: 'pass1234',
    });

    const comment = await request(
      'POST',
      `/api/users/${ada.body.user.id}/comments`,
      { text: 'Absurdly premium.' },
      bob.body.token,
    );
    assert.equal(comment.status, 201);
    assert.equal(comment.body.comment.text, 'Absurdly premium.');

    const report = await request(
      'POST',
      `/api/users/${ada.body.user.id}/comments/${comment.body.comment.id}/report`,
      { reason: 'Testing moderation' },
      ada.body.token,
    );
    assert.equal(report.status, 200);
    assert.equal(report.body.comment.reports.length, 1);

    const reports = await request('GET', '/api/admin/reports', undefined, admin.body.token);
    assert.equal(reports.status, 200);
    assert.equal(reports.body.reports.length, 1);

    const nonAdminReports = await request('GET', '/api/admin/reports', undefined, bob.body.token);
    assert.equal(nonAdminReports.status, 403);
    assert.equal(nonAdminReports.body.code, 'admin_required');

    const dismiss = await request(
      'POST',
      `/api/admin/comments/${comment.body.comment.id}/dismiss-reports`,
      undefined,
      admin.body.token,
    );
    assert.equal(dismiss.status, 200);
    assert.equal(dismiss.body.comment.reports.length, 0);

    const deleteAccount = await request('DELETE', '/api/me', undefined, bob.body.token);
    assert.equal(deleteAccount.status, 200);
    assert.equal(deleteAccount.body.deleted, true);

    const deletedLogin = await request('POST', '/api/auth/login', {
      email: 'bob@example.test',
      password: 'pass1234',
    });
    assert.equal(deletedLogin.status, 403);
    assert.equal(deletedLogin.body.code, 'account_deleted');
  });
});

test('rejects purchases that exceed the daily spending guardrail', async () => {
  await withServer(async ({ request }) => {
    const user = await request('POST', '/api/auth/register', {
      username: 'Ada',
      email: 'ada@example.test',
      password: 'pass1234',
    });

    const first = await request(
      'POST',
      '/api/purchases',
      { expression: '1 + 1', result: '2', amountMinor: 6000 },
      user.body.token,
    );
    assert.equal(first.status, 201);

    const overLimit = await request(
      'POST',
      '/api/purchases',
      { expression: '2 + 2', result: '4', amountMinor: 5000 },
      user.body.token,
    );
    assert.equal(overLimit.status, 429);
    assert.equal(overLimit.body.code, 'daily_spend_limit');
    assert.match(overLimit.body.message, /daily spending limit/i);
  });
});

test('ranks rooms and creator challenges by spend, highest unlock, ridiculousness, and speed', async () => {
  await withServer(async ({ request }) => {
    const ada = await request('POST', '/api/auth/register', {
      username: 'Ada',
      email: 'ada@example.test',
      password: 'pass1234',
    });
    const bob = await request('POST', '/api/auth/register', {
      username: 'Bob',
      email: 'bob@example.test',
      password: 'pass1234',
    });
    const room = await request(
      'POST',
      '/api/rooms',
      { title: 'Ridiculous Math Room' },
      ada.body.token,
    );
    await request(
      'POST',
      '/api/challenges',
      { slug: 'streamer-night', title: 'Streamer Night' },
      ada.body.token,
    );

    const sharedContext = {
      roomCode: room.body.room.code,
      challengeSlug: 'streamer-night',
    };
    await request(
      'POST',
      '/api/purchases',
      {
        expression: '999999999 * (888888 + 7777) / 3 - 1',
        result: '299999666370',
        amountMinor: 900,
        durationMs: 8200,
        context: sharedContext,
      },
      ada.body.token,
    );
    await request(
      'POST',
      '/api/purchases',
      {
        expression: '2 + 2',
        result: '4',
        amountMinor: 500,
        durationMs: 1100,
        context: sharedContext,
      },
      bob.body.token,
    );
    await request(
      'POST',
      '/api/purchases',
      {
        expression: '3 + 3',
        result: '6',
        amountMinor: 500,
        durationMs: 1300,
        context: sharedContext,
      },
      bob.body.token,
    );

    const roomCompetition = await request(
      'GET',
      `/api/rooms/${room.body.room.code}/competition`,
    );
    assert.equal(roomCompetition.status, 200);
    assert.equal(roomCompetition.body.leaders.spent[0].username, 'Bob');
    assert.equal(roomCompetition.body.leaders.highestUnlock[0].username, 'Ada');
    assert.equal(roomCompetition.body.leaders.ridiculous[0].username, 'Ada');
    assert.equal(roomCompetition.body.leaders.fastest[0].username, 'Bob');
    assert.ok(roomCompetition.body.leaders.ridiculous[0].ridiculousScore > 0);
    assert.equal(roomCompetition.body.leaders.fastest[0].fastestRevealMs, 1100);

    const challengeCompetition = await request(
      'GET',
      '/api/challenges/streamer-night/competition',
    );
    assert.equal(challengeCompetition.status, 200);
    assert.equal(challengeCompetition.body.leaders.spent[0].username, 'Bob');
    assert.equal(challengeCompetition.body.leaders.highestUnlock[0].username, 'Ada');
    assert.equal(challengeCompetition.body.leaders.ridiculous[0].username, 'Ada');
    assert.equal(challengeCompetition.body.leaders.fastest[0].username, 'Bob');
  });
});

test('weekly leaderboard scopes purchases to the requested ISO week', async () => {
  let currentNow = '2026-06-21T12:00:00.000Z';
  await withServer(
    async ({ request }) => {
      const ada = await request('POST', '/api/auth/register', {
        username: 'Ada',
        email: 'ada@example.test',
        password: 'pass1234',
      });
      const bob = await request('POST', '/api/auth/register', {
        username: 'Bob',
        email: 'bob@example.test',
        password: 'pass1234',
      });

      await request(
        'POST',
        '/api/purchases',
        { expression: '10 + 1', result: '11', amountMinor: 700 },
        ada.body.token,
      );
      currentNow = '2026-06-22T12:00:00.000Z';
      await request(
        'POST',
        '/api/purchases',
        { expression: '10 + 2', result: '12', amountMinor: 800 },
        bob.body.token,
      );

      const week25 = await request('GET', '/api/weekly/2026-W25/leaderboard');
      assert.equal(week25.status, 200);
      assert.equal(week25.body.weekKey, '2026-W25');
      assert.equal(week25.body.users.length, 1);
      assert.equal(week25.body.users[0].username, 'Ada');

      const week26 = await request('GET', '/api/weekly/2026-W26/leaderboard');
      assert.equal(week26.status, 200);
      assert.equal(week26.body.weekKey, '2026-W26');
      assert.equal(week26.body.users.length, 1);
      assert.equal(week26.body.users[0].username, 'Bob');
    },
    { now: () => new Date(currentNow) },
  );
});
